lib/Doctrine/ORM/UnitOfWork.php

```diff
@@ -1597,6 +1597,26 @@ public function addToIdentityMap($entity)
         $className     = $classMetadata->rootEntityName;
 
         if (isset($this->identityMap[$className][$idHash])) {
+            if ($this->identityMap[$className][$idHash] !== $entity) {
+                throw new RuntimeException(sprintf(
+                    <<<'EXCEPTION'
+While adding an entity of class %s with an ID hash of "%s" to the identity map,
+another object of class %s was already present for the same ID. This exception
+is a safeguard against an internal inconsistency - IDs should uniquely map to
+entity object instances. This problem may occur if you use application-provided
+IDs and reuse ID values. Or, if database-provided IDs are reassigned after
+truncating the database without clearing the EntityManager. Or, you might have
+been using EntityManager#getReference() to create a reference for a nonexistent
+ID that was subsequently (by the RDBMS) assigned to another entity. Otherwise,
+it might be an ORM-internal inconsistency, please report it. 
+EXCEPTION
+                    ,
```

Comment:
if linebreaks are now kept, could also do

```
                    <<<'EXCEPTION'
While adding an entity of class %s with an ID hash of "%s" to the identity map,
another object of class %s was already present for the same ID. This exception
is a safeguard against an internal inconsistency - IDs should uniquely map to
entity object instances. This problem may occur if 

- you use application-provided IDs and reuse ID values.
- database-provided IDs are reassigned after truncating the database without 
  clearing the EntityManager.
- you might have been using EntityManager#getReference() to create a reference 
  for a nonexistent ID that was subsequently (by the RDBMS) assigned to another 
  entity. 

Otherwise, it might be an ORM-internal inconsistency, please report it. 
EXCEPTION
``

Reply:
Nice! I'll use that.